### PR TITLE
rlp: reduce allocations for big.Int and byte array encoding

### PR DESCRIPTION
**Pull request by fjl**
This change further improves the performance of RLP encoding by removing
allocations for `big.Int` and `[...]byte` types. I have added a new benchmark
that measures RLP encoding of `types.Block` to verify that performance is
improved.

Here are the results as measured on my flimsy 1.2GHz Intel Core m3:

```
name           old time/op    new time/op    delta
EncodeBlock-4    68.6µs ± 6%    50.2µs ± 7%  -26.81%  (p=0.000 n=15+15)
name           old alloc/op   new alloc/op   delta
EncodeBlock-4    9.73kB ± 0%    0.06kB ± 1%  -99.34%  (p=0.000 n=11+15)
name           old allocs/op  new allocs/op  delta
EncodeBlock-4       465 ± 0%         1 ± 0%  -99.78%  (p=0.000 n=15+15)
```

### CODE REVIEW

File: rlp/encode.go [hunk 1]
```diff
@@ -397,22 +399,93 @@ func writeBigIntNoPtr(val reflect.Value, w *encbuf) error {
...
 		return fmt.Errorf("rlp: cannot encode negative *big.Int")
-	} else if cmp == 0 {
+	case 0:
 		w.str = append(w.str, 0x80)
-	} else {
-		w.encodeString(i.Bytes())
+	default:
+		bitlen := i.BitLen()
+		if bitlen <= 64 {
+			w.encodeUint(uint64(i.Bits()[0]))
```
Comment (holiman) [inline rlp/encode.go:410]: This won't work on 32 bit architecture, right? Because then the first word is only 32 bits, not 64.

Comment (MariusVanDerWijden) [reply rlp/encode.go:410]: `uint64(i.Int64())` should do the trick right?

File: rlp/encode.go [hunk 2]
```diff
@@ -397,22 +399,93 @@ func writeBigIntNoPtr(val reflect.Value, w *encbuf) error {
...
 	}
 	return nil
 }
 
+// wordBytes is the number of bytes in a big.Word
+const wordBytes = (32 << (uint64(^big.Word(0)) >> 63)) / 8
+
+// writeBigIntBits encodes big.Int values > 2^64.
+func writeBigIntBits(bitlen int, words []big.Word, w *encbuf) {
+	length := ((bitlen + 7) & -8) >> 3
```
Comment (MariusVanDerWijden) [inline rlp/encode.go:423]: could you explain this formula a bit? It's not obvious what it means

Comment (fjl) [reply rlp/encode.go:416]: done

**Comment by holiman** [comment]
Can confirm
```
name           old time/op    new time/op    delta
EncodeBlock-6    63.2µs ±16%    45.2µs ± 5%  -28.53%  (p=0.000 n=10+8)

name           old alloc/op   new alloc/op   delta
EncodeBlock-6    9.73kB ± 0%    0.07kB ± 0%  -99.32%  (p=0.000 n=10+8)

name           old allocs/op  new allocs/op  delta
EncodeBlock-6       465 ± 0%         1 ± 0%  -99.78%  (p=0.000 n=10+10)
```

File: rlp/encode.go [hunk 3]
```diff
@@ -396,13 +394,32 @@ func writeBigIntNoPtr(val reflect.Value, w *encbuf) error {
 	return writeBigInt(&i, w)
 }
 
+// wordBytes is the number of bytes in a big.Word
+const wordBytes = (32 << (uint64(^big.Word(0)) >> 63)) / 8
+
 func writeBigInt(i *big.Int, w *encbuf) error {
-	if cmp := i.Cmp(big0); cmp == -1 {
+	if i.Sign() == -1 {
 		return fmt.Errorf("rlp: cannot encode negative *big.Int")
```
Comment (holiman) [inline rlp/encode.go:402]: Nit: this ought to be an `errors.New("rlp: cannot encode negative *big.Int")`

**Pull request merged**